module/zfs/zio_compress.c

```diff
@@ -145,8 +144,19 @@ zio_compress_data(enum zio_compress c, abd_t *src, void **dst, size_t s_len,
 	if (c == ZIO_COMPRESS_EMPTY)
 		return (s_len);
 
-	/* Compress at least 12.5% */
-	d_len = s_len - (s_len >> 3);
+	/*
+	 * Usable compression thresholds are:
+	 * - less than BPE_PAYLOAD_SIZE (embedded_data feature)
+	 * - at least one saved sector
+	 * - minimun 12.5% should be saved (legacy value)
+	 */
+	if (compress_threshold > 0) {
+		d_len = MAX(BPE_PAYLOAD_SIZE,
+		    s_len - MAX(s_len % compress_threshold, s_len >> 3));
```

Comment:
I think `s_len >> 3` here could also be rounded up or down to `compress_threshold`, otherwise it may produce half-full last sector which either give us no real saving or will be excessively strict when it is not required.

Reply:
I propose to leave `s_len >> 3` as is in this exact patch to make it equal with previous calculations, to minimize threshold changes here.

But it's a great addition for next patches about default threshold changes!

Reply:
Not that I strongly object, but IMO it is not exactly a rocket science to postpone.  I do see that this PR is hanging here for years, but I would prefer it to be done once and done right.  What would you say about this:
```suggestion
                d_len = s_len - (s_len >> 3);
		d_len = MAX(BPE_PAYLOAD_SIZE, d_len - d_len % compress_threshold);
```
On a downside this still does not account for spa_min_alloc, which we do not have here now.  In case lsize/s_len minus 12.5% and rounded down to spa_gcd_alloc is less then spa_min_alloc, we should not even try to compress beyond BPE_PAYLOAD_SIZE, since it will not have sense.  I am thinking if we could move all this math out of here and may be into a separate function alike to zio_roundup_alloc_size() and pass to zio_compress_data() as fully prepared value.

Reply:
@amotin thank you for detailed comment, I moved it into separate logic, now it looks much better indeed! I've used your math, agree with it.